lib/internal/test_runner/snapshot.js

```diff
@@ -0,0 +1,234 @@
+  'the command with the --test-update-snapshots flag.';
+const defaultSerializers = [
+  (value) => { return JSONStringify(value, null, 2); },
+];
+
+function defaultResolveSnapshotPath(testPath) {
+  if (typeof testPath !== 'string') {
+    return testPath;
+  }
+
+  return `${testPath}.snapshot`;
+}
+
+let resolveSnapshotPathFn = defaultResolveSnapshotPath;
+let serializerFns = defaultSerializers;
+
+function setResolveSnapshotPath(fn) {
+  emitExperimentalWarning(kExperimentalWarning);
+  validateFunction(fn, 'fn');
+  resolveSnapshotPathFn = fn;
+}
+
+function setDefaultSerializers(serializers) {
+  emitExperimentalWarning(kExperimentalWarning);
+  validateFunctionArray(serializers, 'serializers');
+  serializerFns = ArrayPrototypeMap(serializers, (fn) => {
+    return fn;
+  });
+}
+
+class SnapshotManager {
+  constructor(entryFile, updateSnapshots) {
+    this.entryFile = entryFile;
+    this.snapshotFile = undefined;
+    this.snapshots = { __proto__: null };
+    this.nameCounts = new SafeMap();
+    // A manager instance will only read or write snapshot files based on the
+    // updateSnapshots argument.
+    this.loaded = updateSnapshots;
+    this.updateSnapshots = updateSnapshots;
+  }
+
+  resolveSnapshotFile() {
+    if (this.snapshotFile === undefined) {
+      const resolved = resolveSnapshotPathFn(this.entryFile);
+
+      if (typeof resolved !== 'string') {
+        const err = new ERR_INVALID_STATE('Invalid snapshot filename.');
+        err.filename = resolved;
+        throw err;
+      }
+
+      this.snapshotFile = resolved;
+    }
+  }
+
+  serialize(input, serializers = serializerFns) {
+    try {
+      let value = input;
+
+      for (let i = 0; i < serializers.length; ++i) {
+        const fn = serializers[i];
+        value = fn(value);
+      }
+
+      return `\n${templateEscape(value)}\n`;
+    } catch (err) {
+      const error = new ERR_INVALID_STATE(
+        'The provided serializers did not generate a string.',
+      );
+      error.input = input;
+      error.cause = err;
+      throw error;
+    }
+  }
+
+  getSnapshot(id) {
+    if (!(id in this.snapshots)) {
+      const err = new ERR_INVALID_STATE(`Snapshot '${id}' not found in ` +
+        `'${this.snapshotFile}.' ${kMissingSnapshotTip}`);
+      err.snapshot = id;
+      err.filename = this.snapshotFile;
+      throw err;
+    }
+
+    return this.snapshots[id];
+  }
+
+  setSnapshot(id, value) {
+    this.snapshots[templateEscape(id)] = value;
+  }
+
+  nextId(name) {
+    const count = this.nameCounts.get(name) ?? 1;
+
+    this.nameCounts.set(name, count + 1);
+    return `${name} ${count}`;
+  }
+
+  readSnapshotFile() {
+    if (this.loaded) {
+      debug('skipping read of snapshot file');
+      return;
+    }
+
+    try {
+      const source = readFileSync(this.snapshotFile, 'utf8');
+      const context = { __proto__: null, exports: { __proto__: null } };
+
+      createContext(context);
+      runInContext(source, context);
```

Comment:
wondering - why `eval` here if it's all JSON/strings instead of JSON.parsing it and saving it as keys on an object instead of js object keys?

Reply:
Readability and debugging. This is also how Jest, Vitest, Deno, node-tap, and probably others store their snapshots. If you serialize as a big JSON object things get very unreadable quickly.